From a9a5901a343311130c5cfbe20c477fa61f2f88ad Mon Sep 17 00:00:00 2001
From: Gerald Elder-Vass <gerald.elder-vass@citrix.com>
Date: Wed, 16 Aug 2023 15:28:48 +0000
Subject: [PATCH] CA-381594: Check ntp_servers before use in DHCP install

---
 backend.py | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/backend.py b/backend.py
index 51d240e..d3b0ad3 100644
--- a/backend.py
+++ b/backend.py
@@ -474,8 +474,10 @@ def rewriteNTPConf(root, ntp_servers):
     ntpsconf = open("%s/etc/chrony.conf" % root, 'w')
     for line in lines:
         ntpsconf.write(line)
-    for server in ntp_servers:
-        ntpsconf.write("server %s iburst\n" % server)
+
+    if ntp_servers:
+        for server in ntp_servers:
+            ntpsconf.write("server %s iburst\n" % server)
     ntpsconf.close()
 
 def setTimeNTP(ntp_servers, ntp_config_method):
-- 
2.39.3

