From fe4f04abcc510f9699348fb88d84289d2e6a4566 Mon Sep 17 00:00:00 2001
From: Yann Dirson <yann.dirson@vates.tech>
Date: Mon, 26 Jan 2026 18:17:31 +0100
Subject: [PATCH 3/3] Make sure Alma/EPEL repositories are disabled by default

---
 backend.py | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/backend.py b/backend.py
index f088574..52d3925 100644
--- a/backend.py
+++ b/backend.py
@@ -1324,6 +1324,12 @@ def configureXcpng(mounts):
     assert util.runCmd2(['chroot', mounts['root'],
                          'systemctl', 'disable', 'iptables.service']) == 0
 
+    # make sure Alma/EPEL repositories are disabled by default
+    assert util.runCmd2(['chroot', mounts['root'],
+                         'dnf', 'config-manager', '--set-disabled', '*']) == 0
+    assert util.runCmd2(['chroot', mounts['root'],
+                         'dnf', 'config-manager', '--set-enabled', 'xcp-ng-*']) == 0
+
     # EL10 disabled ssh root login
     with open(os.path.join(mounts['root'], 'etc/ssh/sshd_config'), "at") as ssh_cfg:
         print("PermitRootLogin yes", file=ssh_cfg)
-- 
2.47.3

